### PR TITLE
Prepare to unsync files to make search / geocoding / poi concurrent

### PR DESCRIPTION
**Pull request by vshcherb**

### CODE REVIEW

**Comment by alisa911** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/1903

**Comment by alisa911** [comment]
3 bbox
Screenshot 2023-07-12 at 16 19 47
1 bbox
Screenshot 2023-07-12 at 16 12 40

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java [hunk 1]
```diff
@@ -126,12 +131,71 @@ public class OsmAndMapsService {
...
+				if (Boolean.TRUE.equals(b)) {
+					cnt++;
+				}
+			}
+			return cnt;
+		}
+		
+		public void unlockReader(BinaryMapIndexReader ret) {
+			if (ret != null && !readers.isEmpty()) {
+				readers.computeIfPresent(ret, (key, value) -> true);
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java:196]: even if it's not present it's still open should be put(ret, true)

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java [hunk 2]
```diff
@@ -771,28 +844,48 @@ private synchronized void initNewObfFiles(File target, File targetTemp) throws I
...
+		ref.fileIndex = cacheFiles.addToCache(reader, target);
 		cacheFiles.writeToFile(new File(config.cacheLocation, CachedOsmandIndexes.INDEXES_DEFAULT_FILENAME));
 		if (nativelib != null) {
 			nativelib.initMapFile(target.getAbsolutePath(), false);
 		}
 		LOGGER.info("Init new obf file " + target.getName() + " " + (System.currentTimeMillis() - val) + " ms");
 	}
 	
-	public synchronized BinaryMapIndexReader[] getObfReaders(QuadRect quadRect, List<LatLon> bbox) throws IOException {
+	public BinaryMapIndexReader[] getObfReaders(QuadRect quadRect, List<LatLon> bbox, int maxNumberMaps) throws IOException, InterruptedException {
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java:869]: this file should return BinaryMapIndexReaderReference, to use pattern try(lock)/finally (unlock)

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/SearchService.java [hunk 3]
```diff
@@ -139,6 +139,7 @@ public synchronized PoiSearchResult searchPoi(SearchService.PoiSearchData data)
             }
             leftoverLimit = limit - res.size();
         }
+        osmAndMapsService.obfFiles.values().forEach(ref -> mapList.forEach(ref::unlockReader));
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/SearchService.java:142]: **Use pattern** should be inside 1 method
try{ 
 lock
} finally {
 unlock
}

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java [hunk 4]
```diff
@@ -852,11 +1011,12 @@ private HashSet<String> getRegionsNameByBbox(List<LatLon> bbox) throws IOExcepti
 			res = osmandRegions.getRegionsToDownload(point.getLatitude(), point.getLongitude(), res);
 			regions.addAll(res);
 		}
+		LOGGER.info("Regions by bbox size " + regions.size());
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java:1014]: not needed in production

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java [hunk 5]
```diff
@@ -126,12 +111,115 @@ public class OsmAndMapsService {
...
+	}
+	
+	@Scheduled(fixedRate = INTERVAL_TO_MONITOR_ZIP)
+	public void closeMapReaders() {
+		obfFiles.forEach((name, ref) -> {
+			if (!ref.readers.isEmpty()) {
+				ref.readers.forEach((reader, open) -> {
+					if (Boolean.TRUE.equals(open)) {
+						try {
+							reader.close();
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java:214]: no synchronization.

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java [hunk 6]
```diff
@@ -126,12 +111,115 @@ public class OsmAndMapsService {
...
+	
+	public void unlockReaders(List<BinaryMapIndexReader> mapsReaders) {
+		obfFiles.values().forEach(ref -> mapsReaders.forEach(ref::unlockReader));
+	}
+	
+	@Scheduled(fixedRate = INTERVAL_TO_MONITOR_ZIP)
+	public void closeMapReaders() {
+		obfFiles.forEach((name, ref) -> {
+			if (!ref.readers.isEmpty()) {
+				ref.readers.forEach((reader, open) -> {
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java:211]: ref.closeUnusedReaders()

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java [hunk 7]
```diff
@@ -126,12 +111,115 @@ public class OsmAndMapsService {
...
-
-	public static class BinaryMapIndexReaderReference {
+	
+	public class BinaryMapIndexReaderReference {
 		File file;
-		BinaryMapIndexReader reader;
+		private static final int WAIT_LOCK_CHECK = 10;
+		ConcurrentHashMap<BinaryMapIndexReader, Boolean> readers = new ConcurrentHashMap<>();
 		public FileIndex fileIndex;
+
```
Comment (vshcherb) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/OsmAndMapsService.java:120]: private synchronized closeUnusedReaders()

**Pull request merged**